### DISCUSSION TITLE
Half-open and invalid message id

### DISCUSSION THREAD

**Discussion by jfhren-stormshield** [strongswan/strongswan#1775] [General]
Hello,

I encountered a problem with the half-open IKE_SA tracking before queuing a job to process an incoming IKE_SA_INIT message introduced in 5.9.6.

When an IKE_SA_INIT request is received, an half-open entry is created when a job is created :
https://github.com/strongswan/strongswan/blob/4c2747fbfc524801cf2d780f3160bca75196fd26/src/libcharon/processing/jobs/process_message_job.c#L137-L142

When the job is executed, an IKE_SA is checked out :
https://github.com/strongswan/strongswan/blob/4c2747fbfc524801cf2d780f3160bca75196fd26/src/libcharon/processing/jobs/process_message_job.c#L67-L68

The message SPI_r and id are checked :
https://github.com/strongswan/strongswan/blob/4c2747fbfc524801cf2d780f3160bca75196fd26/src/libcharon/sa/ike_sa_manager.c#L1329-L1340

If the message id is not 0; the received message is not considered as a true IKE_SA_INIT and `is_init` is not set to `TRUE` leading to the half-open entry not being removed. If repeated, the half-open count as responder is increased triggering the cookie protection if enabled. If the same message (wrong id) but with a correct cookie is received, the half-open count as responder is further increased leading to other protections being triggered like the limitation on half-open IKE_SA if enabled.

Maybe once the job is done, if no IKE_SA was checked out and an half-open entry was created it should be removed ?

**Comment by tobiasbrunner** [comment]
Good catch. I think we should probably just apply the same conditions in both functions. I've pushed a fix to the _1775-init-tracking_ branch.

**Reply by jfhren-stormshield** [reply]
I tested the fix and it works for me. Thank you for your help.

**Reply by tobiasbrunner** [reply]
Thanks for testing, I'll line this up for the next release.
